### ISSUE TITLE
Cardinality aggregation should not reserve a fixed amount of memory per bucket

### ISSUE THREAD

**Issue by jpountz** [elastic/elasticsearch#15892]
The cardinality aggregation currently reserves a fixed amount of memory per bucket. While this makes memory management easier, this also has the downside of making buckets that don't collect many docs very memory-intensive.

**Comment by colings86** [comment]
This has been encountered by a user in https://github.com/elastic/elastisearch/issues/24359#issuecomment-297684702 since the parent terms aggregation was using global ordinals and this was causing a huge number of buckets to be pre-allocated

**Comment by jay-dihenkar** [comment]
+1 : to fix this. ES May go OOM immediately if a user runs this kind of query while testing as mentioned in #24359 and it may be serious if env is PROD.

**Comment by colings86** [comment]
Just to clarify, when the mentioned bug occurs ES trips the circuit breaker and fails the request with a circuit breaker exception rather than ES throwing a OutOfMemoryError and bringing down the node. This is not to say that this isn't a bug or that this bug is less severe, I just want to clarify the behaviour of this bug

**Comment by JeffBolle** [comment]
I must have been doing something wrong, (maybe my circuit breakers were set incorrectly), but when I encountered this the OOM exception brought down ES on the impacted node. #24582

**Comment by jay-dihenkar** [comment]
I would like to confirm here that on versions we tested: v5.2.2 and v5.3.0/1 - ES goes OOM even before the circuit breaker could react.

**Comment by JeffBolle** [comment]
@jay-dihenkar Thanks for confirming, I was running 5.4.0.
@colings86 The circuit breakers do NOT kick in causing just the request to fail.  The thread exits (please reference my logs in #24582), and in my case, Elasticsearch becomes unresponsive on the node. In my case this happened on every node in my cluster, I assume because they all had shards for this query, and it brought my cluster down.

**Comment by colings86** [comment]
@JeffBolle ok noted, I had not seen #24582 and we basing my comment on the stack trace from #24359 which is a circuit breaker exception. So as well as not reserving a fixed amount of memory per bucket in the cardinality aggregation we should also make sure that the cardinality and terms aggregations do a better job of reporting the amount of memory they are about to allocate to the circuit breaker

**Comment by jay-dihenkar** [comment]
I was able to capture the exception in #24359 only after a tweak to circuit breakers limit. Circuit breaker only came into play when I tweaked the settings as follows (after trial and errors): 
```
indices.breaker.total.limit: 50%
indices.breaker.request.limit: 50%
indices.requests.cache.size: 30%
indices.breaker.request.overhead: 550
```
Defaults didn't help prevent the crash.

**Comment by markharwood** [comment]
cc @elastic/es-search-aggs

**Comment by not-napoleon** [comment]
Discussed on 2020-08-19.  We agree this is still relevant, and would like to address it at some point.

**Comment by polyfractal** [comment]
@iverase think we can close this now that https://github.com/elastic/elasticsearch/pull/62560 merged, or still relevant / leave open?

**Comment by nik9000** [comment]
Closed by #62560. Cardinality can still use a bunch of memory but it only once the cardinality grows beyond a certain point.
